File: extensions/interactions/FractionInput/FractionInput.js
// Copyright 2017 The Oppia Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS-IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.



oppia.directive('oppiaInteractiveFractionInput', [function() {
  return {
    restrict: 'E',
    scope: {
      onSubmit: '&'
    },
    templateUrl: 'interaction/FractionInput',
    controller: [
      '$scope', '$attrs', 'FocusManagerService',
      'fractionInputRulesService', 'FractionObjectFactory',
      function($scope, $attrs, FocusManagerService,
        fractionInputRulesService, FractionObjectFactory) {
        $scope.answer = '';
        $scope.labelForFocusTarget = $attrs.labelForFocusTarget || null;
        var requireSimplestForm =
          $attrs.requireSimplestFormWithValue === 'true';
        var errorMessage = '';
        $scope.FRACTION_INPUT_FORM_SCHEMA = {
          type: 'unicode',
          ui_config: {}
        };

        $scope.getWarningText = function() {
          return errorMessage;
        }

        $scope.submitAnswer = function(answer) {
          try {
            var fraction = FractionObjectFactory.fromRawInputString(
              answer);
            if (requireSimplestForm &&
              !angular.equals(fraction, fraction.convertToSimplestForm())
            ) {
              errorMessage =
                'Please enter answer in it\'s simplest form';
            } else {
              $scope.onSubmit({
                answer: fraction,
                rulesService: fractionInputRulesService
              });
            }
          } catch (parsingError) {
            errorMessage = parsingError.message;
          }
        };
      }
    ]
  };
}]);

oppia.directive('oppiaResponseFractionInput', [
  'FractionObjectFactory', 'HtmlEscaperService',
  function(FractionObjectFactory, HtmlEscaperService) {
    return {
      restrict: 'E',
      scope: {},
      templateUrl: 'response/FractionInput',
      controller: ['$scope', '$attrs', function($scope, $attrs) {
        var answer = HtmlEscaperService.escapedJsonToObj($attrs.answer);
        $scope.answer = FractionObjectFactory.fromDict(answer).toString();
      }]
    };
  }
]);

/**
 * Rule evaluation for the fraction input.
 */
oppia.factory('fractionInputRulesService', [
  'FractionObjectFactory',
  function(FractionObjectFactory) {
    var toFloat = function(fractionDict) {
      return FractionObjectFactory.fromDict(fractionDict).toFloat();
    };

    return {
      IsEquivalentTo: function(answer, inputs) {
        return toFloat(answer) === toFloat(inputs.f);
      },
      IsEquivalentToAndInSimplestForm: function(answer, inputs) {
        var simplestForm =
          FractionObjectFactory.fromDict(inputs.f).convertToSimplestForm();
        return toFloat(answer) === toFloat(inputs.f) &&
          angular.equals(answer, simplestForm);
      },
      IsExactlyEqualTo: function(answer, inputs) {
        // Only returns true if both answers are structurally equal.
        return angular.equals(answer, inputs.f);
      },
      IsLessThan: function(answer, inputs) {
        return toFloat(answer) < toFloat(inputs.f);
      },
      IsGreaterThan: function(answer, inputs) {
        return toFloat(answer) > toFloat(inputs.f);
      },
      HasWholeNumberEqualTo: function(answer, inputs) {
        return answer.wholeNumber === inputs.f.wholeNumber;
      },
      HasNumeratorEqualTo: function(answer, inputs) {
        return answer.numerator === inputs.f.numerator;
      },
      HasDenominatorEqualTo: function(answer, inputs) {
        return answer.denominator === inputs.f.denominator;
      },
      HasNoFractionalPart: function(answer) {
        return answer.numerator === 0;
      },
    };
  }
]);

File: extensions/interactions/FractionInput/FractionInput.py
# Copyright 2017 The Oppia Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, softwar
# distributed under the License is distributed on an "AS-IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Python configuration for FractionInput interaction."""

from __future__ import annotations

from extensions.interactions import base

from typing import List, Optional

MYPY = False
if MYPY:  # pragma: no cover
    from extensions import domain


class FractionInput(base.BaseInteraction):
    """Interaction for fraction input."""

    name: str = 'Fraction Input'
    description: str = 'Allows learners to enter integers and fractions.'
    display_mode: str = base.DISPLAY_MODE_INLINE
    is_trainable: bool = False
    _dependency_ids: List[str] = []
    answer_type: str = 'Fraction'
    instructions: Optional[str] = None
    narrow_instructions: Optional[str] = None
    needs_summary: bool = False
    can_have_solution: bool = True
    show_generic_submit_button: bool = True

    _customization_arg_specs: List[domain.CustomizationArgSpecsDict] = [{
        'name': 'requireSimplestForm',
        'description': 'Require the learner\'s answer to be in simplest form',
        'schema': {
            'type': 'bool',
        },
        'default_value': False
    }, {
        'name': 'allowImproperFraction',
        'description': 'Allow improper fractions in the learner\'s answer',
        'schema': {
            'type': 'bool',
        },
        'default_value': True
    }, {
        'name': 'allowNonzeroIntegerPart',
        'description': 'Allow the answer to contain an integer part',
        'schema': {
            'type': 'bool',
        },
        'default_value': True
    }, {
        'name': 'customPlaceholder',
        'description': 'Custom placeholder text (optional)',
        'schema': {
            'type': 'custom',
            'obj_type': 'SubtitledUnicode'
        },
        'default_value': {
            'content_id': None,
            'unicode_str': ''
        }
    }]

    _answer_visualization_specs: List[base.AnswerVisualizationSpecsDict] = [{
        # Table with answer counts for top N answers.
        'id': 'FrequencyTable',
        'options': {
            'column_headers': ['Answer', 'Count'],
            'title': 'Top answers',
        },
        'calculation_id': 'Top10AnswerFrequencies',
        'addressed_info_is_supported': True,
    }]
